Add link traversal option

djdv opened the pull request:
This allows us to follow filesystem links (symlinks, junctions, etc.).
The logic should be good but I need a suggestions on what the flag(s) should be. `traverse-links` is a placeholder.

This is needed in `go-ipfs` so that we can traverse the link and add the target instead of adding the link itself.

Stebalien commented on cli/parse.go:
```diff
@@ -475,6 +480,13 @@ func appendFile(fpath string, argDef *cmdkit.Argument, recursive, hidden bool) (
 	fpath = filepath.ToSlash(filepath.Clean(fpath))
 
+	if travLinks {
+		fpath, err := filepath.EvalSymlinks(fpath)
```
Also, as far as I can tell, this won't conserve the filename (the first argument to NewSerialFile below).

keks replied:
I don't know about the speed, but I agree that the filename probably isn't preserved. @djdv could you add tests that verify that the code works as inteded?

djdv commented:
Sorry for the delay, I had no notification on this for some reason.

@Stebalien, @keks 
Good call on dropping EvalSymlinks.

>--dereference and -L

`cp` seems to dereference ALL links. With this patch, we only traverse the arguments provided, not their children.

I'd be interested in adding `-L` in addition to this, or coalescing the 2 into 1 by adding an optional `depth` parameter if we have the ability to do so.
i.e.
`ipfs add -L` would traverse all links like `cp`
`ipfs add -L=1` would traverse only the base argument, as it does now.
`ipfs add -l=2` would traverse the base root and 1 level under that, symlinks at a depth of 2 will be added AS links.
...
Otherwise we may want to use a different parameter name.
Thoughts on this?

>could you add tests 

For sure, I'll try and come up with some automated tests after we decide on the above^

Stebalien commented:
> cp seems to dereference ALL links. With this patch, we only traverse the arguments provided, not their children.

Ah. For some reason I thought this was traversing all links. This would be the `cp -H` option. Given that, ignore my comment about EvalSymlinks being slow. As top-level files, I believe we *have* to recursively eval the symlinks. I assumed that the `EvalSymlinks` call on `.` evaluated the top-level directory and all the other calls were evaluating symlinks on children.

Although... actually, I believe the fast and correct way to do this is to simply call `Stat` instead of `Lstat` [here](https://github.com/ipfs/go-ipfs-cmds/pull/96/files#diff-3be2b56996dc8a3b09575e16f3cc9ef0R483) with no manual link traversal. If we pass a `stat` object to `NewSerialFile` that says the file is either a directory or a regular file, `NewSerialFile` will treat it as such (and follow any symlinks).

---

I'd add `-L` in a different PR and I wouldn't allow levels. I'd just have `-L` evaluate all symlinks recursively (although deduplicating identical subtrees would be a neat trick).

keks commented:
Since this PR only derefences links in the command line and not those inside the tree, I'd change the flags to `-H` (as in cp, ls, chown, ...) and maybe `--dereference-command-line` (as in ls).

djdv commented:
@Stebalien, @keks 
`-H` seems appropriate but it's used in `ipfs add` already for hidden files. Should we just provide the long form or use something else?

Using Stat over Lstat seems to work the same, except it messes up the progress bar for `ipfs add` for some reason.
Above is Stat, below is when we resolve links manually. Any idea why?
![progress bar](https://user-images.githubusercontent.com/13862850/39269432-cadbd24a-48a0-11e8-9e86-dac282915da8.png)

keks commented:
@djdv 
> Should we just provide the long form or use something else?

I think I'd go for long-form only right now and see if we come up with a good short flag later.

> My assumption is that somewhere we're picking up the size of the link when we should be picking up the size of the target

Yes, that's what I'm thinking, too. The progress bar get's the total file size [here](https://github.com/ipfs/go-ipfs/blob/master/core/commands/add.go#L359), you could log the file size there and see if that's what you expect. Also, using `Printf("%T", sizeFile)` you could find out what kind of file that is to further debug the cause.

---

One minor nitpick: In [`cli/parse.go:478`](https://github.com/ipfs/go-ipfs-cmds/pull/96/files#diff-3be2b56996dc8a3b09575e16f3cc9ef0R478) you have two declarations in a row - I'd use a `var ()` block instead.

djdv commented:
@keks, @Stebalien 
Thanks for that!
The types I'm getting are [serialFiles](https://github.com/ipfs/go-ipfs-cmdkit/blob/c2103d7ae7f889e7329673cc3ba55df8b3863b0f/files/serialfile.go#L16).

I have something working but I'm not sure if I'm approaching this right, I'd like some input on this.

I managed to get accurate sizes by modifying [serialFile.Size()](https://github.com/ipfs/go-ipfs-cmdkit/blob/c2103d7ae7f889e7329673cc3ba55df8b3863b0f/files/serialfile.go#L135) to either deep-resolve the size of symlinks or not, depending on their depth. Currently they're just ignored and sized as 0.
With `-H`, root links get sized, and deeper links do not. If we use `-L` we just ignore the depth limit and size all links.

However, I did this by hard coding a depth limit.
Passing variables through to `serialFile.Size()` would require adding a member to its struct, and changing `NewSerialFile` to accept it.
The current signature is 
```go
func NewSerialFile(name, path string, hidden bool, stat os.FileInfo) (File, error)
```
I would likely need something like this
```go
func NewSerialFile(name, path string, hidden, deref, derefAll bool, stat os.FileInfo) (File, error)
```
I figured if this has to change anyway it may be best to do something like this instead:
```go
type Options struct {
	handleHiddenFiles bool
	resolveRootLinks bool
	resolveAllLinks bool
}
func NewSerialFile(name, path string, options Options, stat os.FileInfo) (File, error)
```
In either case, I could just switch off of them inside of Size() like this
```go
//if file is a symlink {
// -H
if f.options.resolveRootLinks {...conditionalySize(link)...}
// -L
if f.options.resolveAllLinks {...alwaysSize(link)...}
// neither
// just return the size of the link itself
//}
```

parkan commented:
+1 to get some eyes on this again!

djdv commented:
@Stebalien 
I think I understand, however, getting the size is not so much an issue. I had a patch previously that resolved the links inside of size as appropriate, but the problem came in deciding when to resolve.

For a recursive approach, we would need to know inside of `Size()` what flags were passed in, so that we either dereference links or not, (and to what depth if we choose to support that).

I have to revisit this though, the patch I had working before, doesn't anymore, because the types have split since then (Need to change method for `SliceFile.Size()`)

I believe we're going to have that same requirement though, since the only other option is to dereference only the commandline (during initial parsing).

That is to say, I think we can have `-H` without further modifications, but `-L` and/or `-L=int` would likely require `Size()` to be aware of some kind of resolve depth limit, or an "alwaysResolveLinks" type of bool. But in any case, it seems like we'd need to pass something through.

djdv commented:
I pushed a current WIP that depends on these:
branch:[commit]
[go-ipfs/feat/link-traverse](https://github.com/djdv/go-ipfs/tree/feat/link-traverse): [[d37e94aaef74ec85b52d1b59175f8fef027ebf66](https://github.com/djdv/go-ipfs/commit/d37e94aaef74ec85b52d1b59175f8fef027ebf66)]
[go-ipfs-cmdkit/fix/serial-size](https://github.com/djdv/go-ipfs-cmdkit/tree/fix/serial-size): [[f5e723e017a30c3b1ddd0dd4070d37bfa136b1c0](https://github.com/djdv/go-ipfs-cmdkit/commit/f5e723e017a30c3b1ddd0dd4070d37bfa136b1c0)]

But seems to work.

I need to test it more, but we also have to come to a decision on what we want to support, and how to change the NewSerialFile options.
In the WIP commits, I just extended it, but I don't think this is ideal.

djdv commented:
Latest patch works, but is inaccurate on the progressbar for 1 of my test scenarios.
I have a tree like this
```
.
├── mostly-real
│   ├── File1 -> ../all-real/File1
│   ├── File2
│   ├── File3
│   └── File4
├── all-real
│   ├── File1
│   ├── File2
│   ├── File3
│   └── File4
├── sym-to-real -> all-real
└── sym-to-mostly-real -> mostly-real
```

Calling `ipfs add -r --dereference-command-line=2 "sym-to-mostly-real"`, adds as expected.
However, you end up in a situation where the link for File1 is being sized as a link, so your progress goes over 100%. (e.g. 100MB/75MB added).

Semi-related, I may need to add link type-detection up front, in `go-ipfs`.
This call `ipfs add --dereference-command-line=2 "sym-to-mostly-real"` succeeds without the `-r` flag, when it should probably fail.
I'm not actually sure.
"sym-to-mostly-real" isn't a directory, but is added recursively with `--deref...`
I'm not sure if there should be special handling around this. 
Likely, I would check if `--deref` is set, and then check the type and if `-r` was set or not, returning the typical error.

djdv commented:
Everything seems to be in order on this end now. I still need to fix the `-r` issue in `go-ipfs` though.
![link traversal](https://user-images.githubusercontent.com/13862850/45712761-21e6ec00-bb5b-11e8-8833-36240f62d31f.png)

I'll have to rewrite the tests as well, however I'm not sure how to best do that here.
Testing in `go-ipfs` would simply be testing `add` against this tree. But inside cmds itself I'm not too sure.

djdv commented:
The names of flags need to be decided on as well.
To recap: `-H` is in use on `ipfs add` for adding hidden files.

This is my current thought:
Have a `--resolve-links=#` parameter, with `--dereference-command-line` and `-L` being aliases to arguments `resolve-links=1` and `resolve-links=-1` respectively.

Something like `dereference-links` and `link-resolve-depth`, seems fine too if not too verbose.
Open to suggestions on any of these.

Stebalien commented:
I'm not sure depth-limited link traversal is really all that useful, from a user's perspective. I'd really just make these two different options:

1. Something that resolves commandline arguments. (like `ls -H`).
2. Something that resolves all symlinks recursively (like `ls -L`).

The second feature will require an additional option to `NewSerialFile` constructor but I'd do that in a followup patch. Note: we'll *also* have to detect symlink cycles in this case.

Stebalien commented:
Note: to avoid breaking everything, we may want to add a new constructor that takes options using the "functional option" pattern (like we do here: https://github.com/libp2p/go-libp2p-kad-dht/blob/master/opts/options.go). That way, we can stop breaking this every time we need a new option.

djdv commented:
>I'm not sure depth-limited link traversal is really all that useful, from a user's perspective.

I'm wondering this myself.  It's only exposed at the moment for testing, but seemed like it might be useful if people wanted to import complex datasets. However, I don't really know if anyone structures data like this.

In implementation, I don't think it will change much if we omit it. But I have to look at it.
In the meantime, if anyone monitoring the thread has examples of nested levels of symlinks being used for something, please speak on it.

>Note: we'll also have to detect symlink cycles in this case.

I need to double check, but I think Go's `EvalSymlinks` handle's this internally. After enough cycles it should return an error. 

>functional options

👍

djdv commented:
Current status; 
No functional arguments, no tests, both flags seem to work as expected.
`-L` is also in use by `--local` so we have no shortflags.

`--dereference-command-line`
https://github.com/ipfs/go-ipfs-cmds/compare/master...djdv:feat/resolve-links
https://github.com/djdv/go-ipfs/tree/feat/link-traverse

`----dereference-links`
https://github.com/ipfs/go-ipfs-cmds/compare/master...djdv:feat/deep-resolve-links
https://github.com/ipfs/go-ipfs-cmdkit/compare/master...djdv:feat/resolve-links
https://github.com/ipfs/go-ipfs/compare/master...djdv:feat/deep-link-traverse

I'll look at this again later and split up PRs.

Stebalien commented:
@djdv can we try adding a sharness test in go-ipfs for a recursive link (`A/B/C -> A`). I don't think `FollowLinks` will catch this cycle.

djdv commented:
@keks 
Sorry, I kind of hastily dumped that here.

`--dereference-command-line` support in `go-ipfs` would be implemented with these changes:
https://github.com/ipfs/go-ipfs-cmds/compare/master...djdv:feat/resolve-links
https://github.com/djdv/go-ipfs/tree/feat/link-traverse
This resolves arguments only.

`--dereference-links`  support in `go-ipfs` would be implemented with these changes, but will go into a separate PR (it's here just for reference):
https://github.com/ipfs/go-ipfs-cmds/compare/master...djdv:feat/deep-resolve-links
https://github.com/ipfs/go-ipfs-cmdkit/compare/master...djdv:feat/resolve-links
https://github.com/ipfs/go-ipfs/compare/master...djdv:feat/deep-link-traverse
This resolves all links, including arguments and all descendants.

Resolving N levels deep doesn't seem useful, so it's been removed to make the `--dereference-links` implementation a little simpler.

***
I believe this should all be good. But the `--dereference-links` PR should be dependent on functional-options, so that will have to be resolved before it can be merged. Not sure if someone wants to handle this independently or not.

There's also no tests at the moment. The `go test`'s I had previously, are no longer valid.
I've been testing using a [local file tree](http://ipfs.io/ipfs/QmWn4vJepGcAEmT36FDL6ca1RxSXiuca9Ew8pb7RZGKbq4) with `go-ipfs`.
I'm not sure a good way to test this isolated from `go-ipfs`, as the multihash output is what I've been checking.

All in all, I guess we need to layout if there are any more unfulfilled requirements for `--dereference-command-line`.
If not, I can open a PR for it in `go-ipfs` to accompany this one.
Then I'd open a new PR for `--dereference-links` and follow the same pattern.
See what's left to do (functional-options, `go-ipfs` tests, ??), and open a `go-ipfs` PR to add the parameter.

Stebalien commented:
(ping me when there is a PR against go-ipfs and @keks has reviewed it).

keks commented on cli/parse.go:
```diff
@@ -79,6 +78,11 @@ func isRecursive(req *cmds.Request) bool {
+func resolveCommandLine(req *cmds.Request) bool {
+	linkOpt, ok := req.Options[cmds.DerefLong].(bool)
+	return linkOpt && ok
+}
```
`linkOpt` will automatically be `false` if `ok` is `false`. So you could just 
```go
linkOpt, _ := ...
return linkOpt
```

djdv replied:
Shortened that and [another section](https://github.com/ipfs/go-ipfs-cmds/compare/75a3c41ad8827a64cba3bfb4aac7e142d002afcd..69aacbe1b3ebb44cd2b82e3dcba3e7c8fda3b622).

djdv replied:
@Stebalien 
Sorry about that, the semantic difference between casting the return from the map `if req.Options[cmds.DerefLong].(bool)`, and casting the return of a type assertion
`safeVar, _ := req.Options[cmds.DerefLong].(bool)` tripped me up.

djdv commented:
@keks 
Sorry for my extreme delay, a few unexpected events came up. 👀

I rebased this patch, and removed some default bool calls.
I also added tests of this in `go-ipfs` here: https://github.com/ipfs/go-ipfs/pull/5801 
[Relevant test](https://github.com/ipfs/go-ipfs/blob/8f48668bcf75b6a731fd30b7252a24a3251ab119/test/sharness/t0040-add-and-cat.sh#L350-L361)

Do you think this is sufficient?

djdv commented on opts.go:
```diff
@@ -14,10 +14,12 @@ const (
 var OptionStreamChannels = cmdkit.BoolOption(ChanOpt, "Stream channel output")
-var OptionTimeout = cmdkit.StringOption(TimeoutOpt, "set a global timeout on the command")
+var OptionTimeout = cmdkit.StringOption(TimeoutOpt, "Set a global timeout on the command")
+var OptionDerefArgs = cmdkit.BoolOption(DerefLong, "Resolve link arguments, instead of adding links as links")
```
Opinions on "Symlinks supplied in arguments, are dereferenced"?

Stebalien replied:
SGTM, minus, the, comma :smile:.

djdv replied:
Ahh — more of an em dash kind of person I see.
Changed.

Stebalien commented:
I'd *like* to be as consistent with other tools as possible (but really, they're both a mouthful).

(at the end of the day, I don't really care much either way)

Stebalien commented:
Ah, well, in that case, I'd be happy either way.

djdv commented:
Likewise, I have no opinion on this.
Even if we were to break the non-existing convention(s), and use something original.

Options like this seem very script heavy to me, I doubt it will get much interactive (shell) use. So I'm not against verbosity. Tab completion is also a thing anyway.

But there's nothing wrong with brevity either.

kevina commented:
Yeah, unless someone is against it I think it is better to go with the shorter `--dereference-args`.

Sorry for the trouble and being with my review.
